pkg/fqdn/re/re.go

```diff
@@ -0,0 +1,55 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+// Package re provides a simple function to access compile regex objects for
+// the FQDN subsystem.
+package re
+
+import (
+	"fmt"
+	"regexp"
+
+	lru "github.com/golang/groupcache/lru"
+)
+
+// CompileRegex provides a conviennce wrapper around the LRU. It will either
+// return the cached regex object or compile the pattern into a regex object if
+// the LRU hasn't been initialized or it doesn't exist in the cache. If it
+// doesn't exist in the cache, it will be added.
+func CompileRegex(p string) (*regexp.Regexp, error) {
+	if RegexCompileLRU == nil {
+		return regexp.Compile(p)
+	}
+	r, ok := RegexCompileLRU.Get(p)
+	if ok {
+		return r.(*regexp.Regexp), nil
+	}
+	n, err := regexp.Compile(p)
+	if err != nil {
+		return nil, fmt.Errorf("failed to compile regex: %w", err)
+	}
+	RegexCompileLRU.Add(p, n)
```

Comment:
:scream: I somehow missed this. Good catch, I don't know why I didn't check my assumption on that. I assumed that a library like this would include its own locking.

```go
// Cache is an LRU cache. It is not safe for concurrent access.
type Cache struct {
```

Reply:
Also worth noting in particular, that even if we protect the read/write of the LRU, the underlying entries are also being returned. So there's also the question of whether the underlying regex object can be concurrently used by multiple goroutines. (Though I'm guessing maybe we already thought about this in the existing design...?)

Reply:
To answer the last question:

https://pkg.go.dev/regexp#Regexp

> A Regexp is safe for concurrent use by multiple goroutines, except for configuration methods, such as Longest.